builtin/logical/transit/lock_manager.go

```diff
@@ -0,0 +1,337 @@
+package transit
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+const (
+	shared    = false
+	exclusive = true
+)
+
+type lockManager struct {
+	// A lock for each named key
+	locks map[string]*sync.RWMutex
+
+	// A mutex for the map itself
+	lockMutex sync.RWMutex
+
+	// If caching is enabled, the map of name to in-memory policy cache
+	cache map[string]*Policy
+
+	// Used for global locking, and as the cache map mutex
+	globalMutex sync.RWMutex
+}
+
+func newLockManager(cacheDisabled bool) *lockManager {
+	lm := &lockManager{
+		locks: map[string]*sync.RWMutex{},
+	}
+	if !cacheDisabled {
+		lm.cache = map[string]*Policy{}
+	}
+	return lm
+}
+
+func (lm *lockManager) CacheActive() bool {
+	return lm.cache != nil
+}
+
+func (lm *lockManager) LockAll(name string) {
+	lm.globalMutex.Lock()
+	lm.LockPolicy(name, exclusive)
+}
+
+func (lm *lockManager) UnlockAll(name string) {
+	lm.UnlockPolicy(name, exclusive)
+	lm.globalMutex.Unlock()
+}
+
+func (lm *lockManager) LockPolicy(name string, writeLock bool) {
+	lm.lockMutex.RLock()
+	lock := lm.locks[name]
+	if lock != nil {
+		// We want to give this up before locking the lock, but it's safe --
+		// the only time we ever write to a value in this map is the first time
+		// we access the value, so it won't be changing out from under us
+		lm.lockMutex.RUnlock()
+		if writeLock {
+			lock.Lock()
```

Comment:
Why do we give up the lock?  I understand the convention here and locking protocol that assumes this will always be safe (and given those assumptions it is), but the above lines are inherently racy w/o an explanation of why.  Long term I'd feel better about having to lock `lock` before unlocking `lm.lockMutex`, and only if you held both locks exclusively could you remove it from `lm.locks`.  What's the approximate number of entries going into the locks map, and what's their life?